File: rules_to_pcre.cpp
#include "parser_options.hpp"

#include <fstream>
#include <iostream>
#include <map>
#include <string>
#include <sstream>

#include <pcrecpp.h>

#include <boost/bimap.hpp>


// These are the keywords which are not supported. Rules which contain these are skipped.
static std::string unsupportedKeywords[] = {"byte_test", "byte_jump", "byte_extract"};
// This is the array of all the keywords which require separation of corresponding patterns as they are to be matched at different locations
// in the payload.
static std::string separatorKeywords[] = {"http_client_body", "http_cookie", "http_raw_cookie", "http_header", "http_raw_header", "http_method",
                                          "http_uri", "http_raw_uri", "http_stat_code", "http_stat_msg", "pkt_data", "file_data"};
// This is the array of raw versions of all the keywords defined in the above array.
// Some have existing keywords while some don't.
static std::string rawSeparatorKeywords[] = {"", "http_raw_cookie", "http_raw_cookie", "http_raw_header", "http_raw_header", "",
                                             "http_raw_uri", "http_raw_uri", "", "", "", ""};
// The following array is of same length as the keywords array above and each character corresponds to a Snort specific pcre modifier.
// '\0' is used in places where no appropriate modifier exists.
static char separatorModifiers[] = {'P', 'C', 'K', 'H', 'D', 'M',
                                    'U', 'I', 'S', 'Y', '\0', '\0'};

/**
 * This function returns a vector of unsupported keywords.
 */
std::vector<std::string>
getUnsupportedKeywords ()
{
  size_t unsupportedSize = sizeof(unsupportedKeywords) / sizeof(std::string);

  return std::vector<std::string>(unsupportedKeywords, unsupportedKeywords + unsupportedSize);
}

std::map<std::string, std::string>
getRawKeywordsMap ()
{
  size_t kwSize = sizeof(rawSeparatorKeywords) / sizeof(std::string);

  std::map<std::string, std::string> rm;
  for (size_t i = 0; i < kwSize; ++i) {
    if (!rawSeparatorKeywords[i].empty()) {
      rm.insert(std::make_pair(rawSeparatorKeywords[i], separatorKeywords[i]));
    }
  }
  return rm;
}

/**
 * This function returns a bi-directional map from all the separatorKeywords to their
 * corresponding indices. This could have returned a vector but returning a bi-directional
 * map is, arguably, more convenient. Later on, that is.
 */
boost::bimap<std::string, size_t>
getSeparatorKeywordIndices ()
{
  size_t kwSize = sizeof(separatorKeywords) / sizeof(std::string);
  std::map<std::string, std::string> rkw(getRawKeywordsMap());

  boost::bimap<std::string, size_t> bm;
  size_t index = 1;
  for (size_t i = 0; i < kwSize; ++i) {
    if (rkw.find(separatorKeywords[i]) == rkw.end()) {
      bm.insert(boost::bimap<std::string, size_t>::value_type(separatorKeywords[i], index++));
    }
  }

  return bm; 
}

/**
 * Indices of Snort specific modifiers in final separated pattern list.
 * This is required for pcre mentioned in rules. 0 is for complete payload.
 */
std::map<char, size_t>
getSeparatorModifierIndices ()
{
  static size_t modSize = sizeof(separatorModifiers) / sizeof(char);

  static std::map<char, size_t> sm;
  for (size_t i = 0; i < modSize; ++i) {
    if (separatorModifiers[i] != '\0') {
      sm.insert(std::make_pair(separatorModifiers[i], i + 1));
    }
  }
  return sm;
}

/**
 * This function reads all the uncommented rules from given rules
 * files and returns options for the rules.
 */
std::vector<std::string>
parseRulesFiles (const std::vector<std::string>& rulesFiles)
{
  std::vector<std::string> allOptions;
  pcrecpp::RE optionPattern("\\((.*(?:content:|pcre:).*)\\)");

  std::vector<std::string> unsupportedKeywords(getUnsupportedKeywords());
  std::string unsupportedString = "(";
  for (std::vector<std::string>::const_iterator uk = unsupportedKeywords.begin(); uk != unsupportedKeywords.end(); ++uk) {
    unsupportedString += (*uk + "|");
  }
  *(unsupportedString.rbegin()) = ')';
  pcrecpp::RE unsupportedPattern(unsupportedString);

  for (std::vector<std::string>::const_iterator rf = rulesFiles.begin(); rf != rulesFiles.end(); ++rf) {
    std::ifstream rfstream((*rf).c_str());
    for (std::string rule; std::getline(rfstream, rule); ) {
      std::string newOption;
      if ((rule[0] != '#') && optionPattern.PartialMatch(rule, &newOption)) {
        //std::cout << newOption << std::endl;
        std::string keyword;
        if (unsupportedPattern.PartialMatch(newOption, &keyword)) {
          std::cerr << std::endl;
          std::cerr << "Keyword \"" << keyword << "\" is not supported. Skipping following rule." << std::endl;
          std::cerr << rule << std::endl << std::endl;
        }
        else {
          allOptions.push_back(newOption);
        }
      }
    }
  }
  return allOptions;
}

/**
 * This function returns a pattern for a given vector of pattern specifiers.
 * The pattern specifiers can be a combination of 'content's and 'pcre's.
 */
std::string
getContentPattern (const std::vector<std::string>& patternVector)
{
  pcrecpp::RE contentPattern("content:(!?)\"(.*)\"");
  pcrecpp::RE contentParamPattern("(offset|depth|distance|within):(\\d+)");
  pcrecpp::RE pcrePattern("pcre:(!?)\"\\/(.*)[\\/]?(\\w*)\"", pcrecpp::RE_Options(PCRE_UNGREEDY));
  pcrecpp::RE escapePattern("(\\.|\\^|\\$|\\*|\\+|\\?|\\(|\\)|\\[|\\{|\\\\|\\/)");
  pcrecpp::RE pipePattern("(.*)\\|((?:[A-F\\d]{2} ?)*)\\|");
  pcrecpp::RE hexPattern("([\\dA-F]{2}) ?");

  std::vector<std::string> independentPatterns;
  for (std::vector<std::string>::const_iterator pattern = patternVector.begin(); pattern != patternVector.end(); ++pattern) {
    std::string negation, thisPattern, thisModifiers;

    bool relativePattern = false;

    size_t escapePatternCount = 0;
    size_t hexPatternCount = 0;

    pcrecpp::StringPiece pString(*pattern);
    if ((*pattern).compare(0, 7, "content") == 0) {
      std::string contentString;
      size_t offset = 0, depth = std::string::npos;
      if (contentPattern.Consume(&pString, &negation, &contentString)) {
        escapePatternCount = escapePattern.GlobalReplace("\\\\\\1", &contentString);
        if (pString.as_string().find("nocase;") != std::string::npos) {
          thisModifiers = "i";
        }

        pcrecpp::StringPiece contentSP(contentString.c_str());
        std::string prefix, isRaw;
        while (pipePattern.FindAndConsume(&contentSP, &prefix, &isRaw)) {
          hexPatternCount += hexPattern.GlobalReplace("\\\\x\\1", &isRaw);
          prefix += (isRaw + contentSP.as_string());
          contentSP.set(prefix.c_str());
        }
        contentString = contentSP.as_string();

        std::string param;
        int value;
        while (contentParamPattern.FindAndConsume(&pString, &param, &value)) {
          if (value < 0) {
            throw std::runtime_error("Handling of negative parameter values is not implemented!");
          }

          if (param == "offset") {
            offset = value;
          }
          else if (param == "depth") {
            depth = value;
          }
          else if (param == "distance") {
            offset = value;
            relativePattern = true;
          }
          else { // param == "within"
            depth = value;
            relativePattern = true;
          }
        }
      }
      else {
        throw std::runtime_error("Provided content pattern didn't match the standard pattern!");
      }
      std::stringstream ps;
      if (!thisModifiers.empty()) {
        ps << "(?" << thisModifiers << ":";
      }
      if ((offset > 0) || (depth < std::string::npos)) {
        size_t contentSize = contentString.length() - (escapePatternCount * 1) - (hexPatternCount * 3);
        if (depth < contentSize) {
          throw std::runtime_error("Encountered depth/within less than content string length!");
        }
        if (!relativePattern) {
          ps << "^";
        }
        size_t end = (depth != std::string::npos) ? (offset + depth) - contentSize : 0;
        if ((offset > 0) || (end > offset)) {
          ps << ".{" << offset;
          if (end > offset) {
            ps << "," << end;
          }
          ps << "}";
        }
        if (depth == std::string::npos) {
          ps << ".*";
        }
      }
      else if (relativePattern) {
        ps << ".*";
      }
      ps << contentString;
      if (!thisModifiers.empty()) {
        ps << ")";
      }
      thisPattern = ps.str();
    }
    else { // a pcre pattern
      if (pcrePattern.Consume(&pString, &negation, &thisPattern, &thisModifiers)) {
        size_t index = thisModifiers.find('R');
        if (index != std::string::npos) {
          thisModifiers.replace(index, 1, "");
          relativePattern = true;
        }
        if (!thisModifiers.empty()) {
          thisPattern = "(?" + thisModifiers + ":" + thisPattern + ")";
        }
      }
      else {
        throw std::runtime_error("Provided pcre pattern didn't match the standard pattern!");
      }
    }
    if (!negation.empty()) {
      thisPattern = "(?!" + thisPattern + ")";
    }
    if (relativePattern && (independentPatterns.size() > 0)) {
      (*(independentPatterns.rbegin())).append(thisPattern);
    }
    else {
      independentPatterns.push_back(thisPattern);
    }
  }

  std::string patternString;
  size_t numPatterns = independentPatterns.size();
  if (numPatterns > 1) {
    for (size_t p = 0; p < (numPatterns - 1); ++p) {
      patternString += "(?=.*" + independentPatterns[p] + ")";
    }
    patternString += ".*";
  }
  patternString += independentPatterns[numPatterns - 1];

  return patternString;
}

/**
 * Main function.
 */
int
main (int argc, char** argv)
{
  ParserOptions po;
  try {
    po.parse(argc, argv);
  }
  catch (po::error& pe) {
    std::cerr << pe.what();
    return 1;
  }
  std::vector<std::string> allOptions = parseRulesFiles(po.rulesFiles());

  pcrecpp::RE sidPattern("sid:(\\d+);");
  pcrecpp::RE contentPattern("((content|pcre):.*)(content:|pcre:|$)", pcrecpp::RE_Options(PCRE_UNGREEDY)); 

  size_t kwSize = sizeof(separatorKeywords) / sizeof(std::string);
  std::string separatorString = "(";
  for (size_t i = 0; i < kwSize; ++i) {
    separatorString += (separatorKeywords[i] + "|");
  }
  *(separatorString.rbegin()) = ')';
  pcrecpp::RE keywordPattern(separatorString);

  std::map<char, size_t> separatorModifiers(getSeparatorModifierIndices());
  separatorString = "(";
  for (std::map<char, size_t>::const_iterator sm = separatorModifiers.begin(); sm != separatorModifiers.end(); ++sm) {
    separatorString += (std::string(1, sm->first) + "|");
  }
  *(separatorString.rbegin()) = ')';
  separatorString = "(pcre:\"\\/.*\\/\\w*)" + separatorString + "(\\w*\")";
  pcrecpp::RE pcrePattern(separatorString);

  std::map<std::string, std::ofstream*> fileMap;
  for (std::vector<std::string>::const_iterator option = allOptions.begin(); option != allOptions.end(); ++option) {
    size_t sid;
    if (!sidPattern.PartialMatch(*option, &sid)) {
      throw std::runtime_error("Encountered a rule with no SID!");
    }
    pcrecpp::StringPiece optionString(*option);
    std::map<std::string, std::string> rawKeywordsMap(getRawKeywordsMap());
    boost::bimap<std::string, size_t> separatorKeywordIndices(getSeparatorKeywordIndices());
    std::string thisContent, type, nextContent;
    std::map<std::pair<size_t, bool>, std::vector<std::string> > contentVectors;
    while (!optionString.empty() && contentPattern.FindAndConsume(&optionString, &thisContent, &type, &nextContent)) {
      size_t index = 0;
      if (type == "content") {
        std::string keyword;
        bool isRaw = false;
        if (keywordPattern.PartialMatch(thisContent, &keyword)) {
          std::map<std::string, std::string>::const_iterator rk = rawKeywordsMap.find(keyword);
          if (rk != rawKeywordsMap.end()) {
            keyword = rk->second;
            isRaw = true;
          }
          index = (separatorKeywordIndices.left.find(keyword))->second;
        }
        if ((*option).find("rawbytes;") != std::string::npos) {
          isRaw = true;
        }
        contentVectors[std::make_pair(index, isRaw)].push_back(thisContent);
      }
      else { // type == "pcre"
        pcrecpp::StringPiece contentString(thisContent);
        std::string pcreString, modifier, suffix;
        if (pcrePattern.Consume(&contentString, &pcreString, &modifier, &suffix)) {
          // Modifier was found in the string
          // Set the index as per the modifier and remove the modifier
          index = (separatorModifiers.find(modifier[0]))->second;
          pcreString += suffix;
          pcreString += contentString.as_string();
        }
        else {
          pcreString = thisContent; 
        }
        contentVectors[std::make_pair(index, false)].push_back(pcreString);
      }
      nextContent += optionString.as_string();
      optionString.set(nextContent.c_str());
    }

    for (std::map<std::pair<size_t, bool>, std::vector<std::string> >::const_iterator content = contentVectors.begin(); content != contentVectors.end(); ++content) {
      try {
        std::string patternString = getContentPattern(content->second);

        std::string outputFile = "payload";
        if ((content->first).first > 0) {
          outputFile = (separatorKeywordIndices.right.find((content->first).first))->second;
        }
        if ((content->first).second) {
          outputFile += "_raw";
        }

        std::ostream* out = 0;
        if (po.writeFiles()) {
          if (fileMap.find(outputFile) == fileMap.end()) {
            std::string fileName = outputFile + ".pcort";
            fileMap[outputFile] = new std::ofstream(fileName.c_str(), std::ofstream::out);
          }
          out = fileMap[outputFile];
        }
        else {
          out = &std::cout;
        }
        *(out) << sid << ": " << patternString << std::endl;
      }
      catch (std::runtime_error& e) {
        std::cerr << std::endl;
        std::cerr << "Getting pattern for rule with SID " << sid << " failed." << std::endl;
        std::cerr << e.what() << std::endl << std::endl;
      }
    }
  }
  for (std::map<std::string, std::ofstream*>::iterator f = fileMap.begin(); f != fileMap.end(); ++f) {
    delete f->second;
    f->second = 0;
  }
  return 0;
}

File: rules_to_pcre.py
from argparse import ArgumentParser, ArgumentTypeError
import os
import re
from sys import stderr
from collections import defaultdict


class RulesConverter(object):
    """
    Class for converting Snort rules to ANML. 
    """
    # list of Snort keywords that are not supported
    _unsupportedKeywords = (
        'byte_test',
        'byte_jump',
        'byte_extract',
    )

    # map of modifier keywords to the corresponding bucket and PCRE keywords
    _keywordsMap = {
        'http_client_body' : ('', 'P'),
        'http_cookie'      : ('', 'C'),
        'http_raw_cookie'  : ('http_cookie', 'K'),
        'http_header'      : ('', 'H'),
        'http_raw_header'  : ('http_header', 'D'),
        'http_method'      : ('', 'M'),
        'http_uri'         : ('', 'U'),
        'http_raw_uri'     : ('http_uri', 'I'),
        'http_stat_code'   : ('', 'S'),
        'http_stat_msg'    : ('', 'Y'),
        'pkt_data'         : ('', ''),
        'file_data'        : ('', ''),
    }

    # compiled patterns for matching and extracting patterns from rules
    _optionPattern = re.compile(r'\((?P<options>.* (?:content|pcre):.*)\)')
    _unsupportedPattern = re.compile(r'(?P<unsupported>%s)'%('|'.join(_unsupportedKeywords)))
    _sidPattern = re.compile(r'sid:(?P<sid>\d+);')
    _genericPattern = re.compile(r'(?P<content>(?P<type>content|pcre):.*?)(?=content:|pcre:|$)')
    _genericPcrePattern = re.compile(r'(?P<pcre>pcre:"/.*/\w*)(?P<modifier>%s)(?P<suffix>\w*")'%('|'.join(sm[-1] for sm in _keywordsMap.itervalues() if sm[-1] != '')))
    _keywordsPattern = re.compile(r'(?P<keyword>%s);'%('|'.join(_keywordsMap.iterkeys())))
    _contentPattern = re.compile(r'content:(?P<negation>!?)"(?P<string>.*)"')
    _paramPattern = re.compile(r'(?P<name>offset|depth|distance|within):(?P<value>\d+)')
    _pcrePattern = re.compile(r'pcre:(?P<negation>!?)"/(?P<pattern>.*?)[/]?(?P<modifiers>\w*)"')
    _escapePattern = re.compile(r'(\.|\^|\$|\*|\+|\?|\(|\)|\{|\[|\\|/)')
    _lookaheadPattern = re.compile('(\(\?=.*\))')
    _pipePattern = re.compile(r'(\|(?P<suffix>(?:[A-F\d]{2} ?)*)\|)')

    # cached modifier keyword map
    _modifierKeywordsMap = None

    @classmethod
    def _get_modifier_keyword(cls, modifier):
        """
        Creates map from PCRE modifier to the corresponding keyword,
        if not already created, and returns the map.
        """
        if cls._modifierKeywordsMap is None:
            cls._modifierKeywordsMap = {}
            for keyword, value in cls._keywordsMap.iteritems():
                if value[-1]:
                    cls._modifierKeywordsMap[value[-1]] = keyword
        return cls._modifierKeywordsMap[modifier]

    @classmethod
    def _get_file_options(cls, rulesFile):
        """
        Extracts all the supported options from the given rules file.
        The argument should be a file object.
        """
        fileOptions = []
        for rule in rulesFile:
            rule = rule.strip()
            if not rule or rule[0] == '#':
                # skip commented rules, denoted by '#'
                # also skip empty lines
                continue
            matched = cls._optionPattern.search(rule)
            if matched is not None:
                newOption = matched.group('options')
                matched = cls._unsupportedPattern.search(newOption)
                if matched is not None:
                    stderr.write('Skipping the following rule as the keyword "%s" is not supported.\n%s\n\n'%(matched.group('unsupported'), rule))
                    stderr.flush()
                else:
                    fileOptions.append(newOption)
            else:
                stderr.write("Skipping the following rule as it doesn't have any pattern matching keywords.\n%s\n\n"%(rule))
                stderr.flush()
        return fileOptions

    @classmethod
    def _get_all_options(cls, rulesFiles):
        """
        Gets all the supported options from the rules file(s).
        """
        allOptions = []
        for f in rulesFiles:
            with open(f, 'rb') as rulesFile:
                allOptions.extend(cls._get_file_options(rulesFile))
        return allOptions

    def __init__(self, writeFiles, maxLookaheads, handleNegations):
        """
        Constructor. Stores some of the program options.
        """
        self._writeFiles = writeFiles
        self._maxLookaheads = maxLookaheads
        self._handleNegations = handleNegations

    def _combine_independent_patterns(self, independentPatterns):
        """
        Combines independent patterns provided as a list.
        """
        patternString = ''
        numPatterns = len(independentPatterns)
        if numPatterns > 1:
            for p in xrange(0, numPatterns - 1):
                if independentPatterns[p][0] == '^':
                    if independentPatterns[-1][0] != '^':
                        temp = independentPatterns[-1]
                        independentPatterns[-1] = independentPatterns[p]
                        independentPatterns[p] = temp
                else:
                    independentPatterns[p] = '.*' + independentPatterns[p]
                patternString += '(?=%s)'%(independentPatterns[p])
            if independentPatterns[-1][0] != '^':
                patternString += '.*'
        patternString += independentPatterns[-1]
        return patternString

    def _get_content_pattern(self, patterns):
        """
        Extracts patterns from given content/pcre. 
        """
        independentPatterns = []
        numLookaheads = 0
        for p in patterns:
            relative = False
            thisModifiers = ''
            thisPattern = ''
            negation = ''

            if p.startswith('content'):
                content = self._contentPattern.search(p)
                if content is not None:
                    offset = 0
                    depth = -1
                    negation = content.group('negation')
                    contentString, escapePatternCount = self._escapePattern.subn(lambda m: '\\' + m.group(1), content.group('string'))
                    class PipeSubFunc(object):
                        _hexPattern = re.compile(r'([\dA-F]{2}) ?')
                        def __init__(self):
                            self.hexPatternCount = 0
                        def __call__(self, m):
                            subString, subCount = self._hexPattern.subn(lambda m : r'\x' + m.group(1), m.group('suffix'))
                            self.hexPatternCount += subCount
                            return subString

                    pipeSubFunc = PipeSubFunc()
                    contentString, subCount = self._pipePattern.subn(pipeSubFunc, contentString)
                    if p.find('nocase;') != -1:
                        thisModifiers = 'i'
                    for param in self._paramPattern.finditer(p):
                        name = param.group('name')
                        value = int(param.group('value'))
                        if value < 0:
                            raise RuntimeError, 'Handling of negative parameter values is not implemented!'
                        offset = value if name in ['offset', 'distance'] else offset
                        depth = value if name in ['depth', 'within'] else depth
                        relative = True if name in ['distance', 'within'] else relative
                    ps = []
                    if offset != 0 or depth != -1:
                        contentSize = len(contentString) - escapePatternCount - (pipeSubFunc.hexPatternCount * 3)
                        if depth != -1 and depth < contentSize:
                            raise RuntimeError, 'Encountered depth/within less than content string length!'
                        if not relative:
                            ps.append('^')
                        end = (offset + depth) - contentSize if depth != -1 else 0
                        if offset > 0 or end > offset:
                            ps.append('.{%d'%(offset))
                            if end > offset:
                                ps.append(',%d'%(end))
                            ps.append('}')
                        if depth == -1:
                            ps.append('.*')
                    elif relative:
                        ps.append('.*')
                    ps.append(contentString)
                    thisPattern = ''.join(ps)
                else:
                    raise RuntimeError, "Provided content pattern didn't match the standard pattern!"
            else:
                matched = self._pcrePattern.search(p)
                if matched is not None:
                    negation = matched.group('negation')
                    thisModifiers = matched.group('modifiers')
                    if thisModifiers.find('R') != -1:
                        thisModifiers = thisModifiers.replace('R', '')
                        relative = True
                    if thisModifiers.find('B') != -1:
                        print p
                    thisPattern = matched.group('pattern')
                    numLookaheads += self._lookaheadPattern.subn('', thisPattern)[1]
                else:
                    raise RuntimeError, "Provided pcre pattern didn't match the standard pattern!"
            if thisModifiers:
                thisPattern = '(?%s:%s)'%(thisModifiers, thisPattern)
            if negation:
                if not self._handleNegations:
                    raise RuntimeError, "Can't handle negations!"
            if relative and len(independentPatterns) > 0:
                independentPatterns[-1] = independentPatterns[-1] + thisPattern
            else:
                independentPatterns.append(thisPattern)
        numLookaheads += (len(independentPatterns) - 1)
        if numLookaheads > self._maxLookaheads:
            raise RuntimeError, 'Number of lookaheads exceeded the maximum limit!'
        return self._combine_independent_patterns(independentPatterns)

    def convert(self, rulesFiles):
        """
        Convert all the rules in given rules files to the corresponding ANML or PCRE. 
        """
        outputFiles = {}
        allSids = []
        supportedSids = []
        for option in self._get_all_options(rulesFiles):
            matched = self._sidPattern.search(option)
            if matched is None:
                raise RuntimeError, "Encountered a rule with no SID!"
            sid = int(matched.group('sid'))
            allSids.append(sid)
            contentVectors = defaultdict(list)
            for pattern in self._genericPattern.finditer(option):
                keyword = 'general'
                raw = False
                thisContent = pattern.group('content')
                if pattern.group('type') == 'content':
                    matched = self._keywordsPattern.search(thisContent)
                    if matched is not None:
                        keyword = matched.group('keyword')
                else:
                    matched = self._genericPcrePattern.search(thisContent)
                    if matched is not None:
                        pcreString = matched.group('pcre') + matched.group('suffix')
                        contentString = self._genericPcrePattern.sub('', thisContent, count = 1)
                        thisContent = pcreString + contentString
                        keyword = self._get_modifier_keyword(matched.group('modifier'))
                raw = option.find('rawbytes;') != -1
                if keyword in self._keywordsMap and self._keywordsMap[keyword][0]:
                    raw = raw or bool(self._keywordsMap[keyword][0])
                    keyword = self._keywordsMap[keyword][0]
                contentVectors[(keyword, raw)].append(thisContent)
            convertedStrings = {}
            handled = True
            for bucket, patterns in contentVectors.iteritems():
                try:
                    convertedStrings[bucket] = self._get_content_pattern(patterns)
                except RuntimeError, e:
                    stderr.write('\nGetting pattern for rule with SID %d failed.\n%s\n\n'%(sid, str(e)))
                    handled = False
                    break
            if not handled:
                continue
            supportedSids.append(sid)
            for keyword, pattern in convertedStrings.iteritems():
                keyword = keyword[0] + '_raw' if keyword[1] else keyword[0]
                writeString = '%d: /%s/'%(sid, pattern)
                if self._writeFiles and keyword not in outputFiles:
                    outputFiles[keyword] = open(keyword + '.txt', 'wb')
                if self._writeFiles:
                    outputFiles[keyword].write(writeString + '\n')
                else:
                    print writeString

if __name__ == '__main__':
    def RulesPath(path):
        allFiles = []
        if os.path.isdir(path):
            for subdirs, dirs, files in os.walk(path):
                allFiles.extend(os.path.join(path, name) for name in files if name.endswith('.rules'))
        elif os.path.isfile(path):
            allFiles.append(path)
        else:
            raise ArgumentTypeError, 'The provided path is neither a file nor a directory!'
        return allFiles

    parser = ArgumentParser()
    parser.add_argument('--rules', help = 'Specify the directory/file from which rules are to be extracted.',
                        required = True, type = RulesPath)
    parser.add_argument('--writefiles', help = 'Flag for specifying if the output should be written to a set of files.',
                        action = 'store_true')
    parser.add_argument('--maxlookaheads', help = 'Maximum number of lookaheads in the generated PCRE patterns',
                        metavar = 'COUNT', type = int, default = 0)
    parser.add_argument('--negations', help = 'Flag for specifying if negations should be handled.',
                        action = 'store_true')
    args = parser.parse_args()

    converter = RulesConverter(args.writefiles, args.maxlookaheads, args.negations)
    converter.convert(args.rules)
